File: analysis/sensitivity.py
from fractions import Fraction
from reference_score import exact_score, half_up_tenths

SCENARIOS = {"A": ({"critical": 40, "serious": 30, "moderate": 20, "minor": 10}, 40), "B": ({"critical": 50, "serious": 30, "moderate": 20, "minor": 10}, 50), "C": ({"critical": 40, "serious": 25, "moderate": 15, "minor": 10}, 40)}

def sensitivity(sites):
    output = {}
    for name, (weights, maximum) in SCENARIOS.items():
        scores = {site: exact_score(rows, weights, maximum) for site, rows in sites.items()}
        scores = {site: value for site, value in scores.items() if value is not None}
        output[name] = {"scores": {site: {"exact": f"{value.numerator}/{value.denominator}", "display": half_up_tenths(value)} for site, value in scores.items()}, "ranking": sorted(scores, key=lambda site: (-scores[site], site))}
    return output

File: src/lib/sensitivity.ts
import {
  DEFAULT_WEIGHTS,
  exactScore,
  MAX_WEIGHT,
  roundHalfUpTenths,
  type ScoreOpportunity,
} from "./score";
import { compareExact } from "./score";
export const SENSITIVITY_SCENARIOS = {
  A: DEFAULT_WEIGHTS,
  B: { critical: 50, serious: 30, moderate: 20, minor: 10 },
  C: { critical: 40, serious: 25, moderate: 15, minor: 10 },
} as const;
export function sensitivityAnalysis(sites: Record<string, ScoreOpportunity[]>) {
  const outputs = Object.entries(SENSITIVITY_SCENARIOS).map(([name, weights]) => {
    const exact: Record<string, NonNullable<ReturnType<typeof exactScore>>> = {};
    for (const [site, opportunities] of Object.entries(sites)) {
      const value = exactScore(opportunities, weights, name === "B" ? 50 : MAX_WEIGHT);
      if (value) exact[site] = value;
    }
    const ranking = Object.keys(exact).sort(
      (a, b) => compareExact(exact[b], exact[a]) || a.localeCompare(b),
    );
    return {
      scenario: name,
      scores: Object.fromEntries(
        Object.entries(exact).map(([site, value]) => [
          site,
          {
            exactNumerator: value.numerator.toString(),
            exactDenominator: value.denominator.toString(),
            display: roundHalfUpTenths(value),
          },
        ]),
      ),
      ranking,
    };
  });
  return outputs;
}
